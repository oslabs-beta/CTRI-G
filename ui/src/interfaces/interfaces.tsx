interface ContainerInfo {
  Name: string;
  Id: string;
  Image: string;
  State: string;
  Networks: string | [string];
  Ports: PortItem[];
}

interface NetworkContainerInfo {
  Name: string;
  IPv4Address: string;
}

interface PortItem {
  IP?: string;
  PrivatePort?: string;
  PublicPort?: string;
  Type?: string;
}

interface NetworkInfo {
  Driver: string;
  Name: string;
  ID: string;
  Containers?: string[];
  IPv4Address?: string[];
  Gateway?: string;
  Subnet?: string;
  Scope?: string;
}


interface forceProps {
  name: string;
  value?: number;
  children?: forceProps[];
}

interface treeProps {
  name: string;
  value?: number;
  children?: treeProps[];
}

interface nodeData {
  from: string;
  to: string;
  value: number;
}

interface graphData extends Array<nodeData> {}

type setNetworks = React.Dispatch<React.SetStateAction<[] | NetworkInfo[]>>;

type setContainers = React.Dispatch<React.SetStateAction<[] | ContainerInfo[]>>;

export type {
  ContainerInfo,
  PortItem,
  NetworkInfo,
  setNetworks,
  setContainers,
  NetworkContainerInfo,
  graphData,
  forceProps,
  treeProps
};
