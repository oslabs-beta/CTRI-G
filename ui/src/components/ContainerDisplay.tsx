import React from 'react';
import {
  ContainerInfo,
  setContainers,
  setNetworks,
  isLoading
} from '../interfaces/interfaces';
import { DisconnectContainer } from '../functions/functions';

//Component to display Container
const ContainerDisplay: React.FC<{
  id: string;
  info: ContainerInfo;
  network: string;
  setContainers: setContainers;
  setNetworks: setNetworks;
  isLoading: isLoading
}> = props => {
  //if the Ports object exist within our props
  if (props.info.Ports) {
    // return container information including Ports info.

    return (
      <div id={props.id} className='container'>
        {/* Display container information*/}
        <div className='containerInfo'>
          <p>
            {' '}
            <strong>Name: </strong>
            <br /> {props.info.Name}
          </p>
          <hr />
          <p>
            <strong>ContainerID: </strong>
            <br /> {props.info.Id}
          </p>
          <hr />
          <p>
            <strong>Image: </strong>
            <br /> {props.info.Image}
          </p>
          <hr />
          <p>
            <strong>Activity: </strong>
            <br /> {props.info.State}
          </p>
          <div className='containerButtons'>
            {/* <button className= 'button' onClick={() => ConnectContainers(props.info.Name) } >Connect</button>  */}
            <button
              className='innerButton'
              onClick={() =>
                DisconnectContainer(
                  props.info.Name,
                  props.network,
                  props.setContainers,
                  props.setNetworks,
                  props.isLoading
                )
              }>
              Disconnect
            </button>
          </div>
        </div>
        <ul className='portInfo'>
          {/* Display list of information from Ports*/}
          <li>
            <strong>IP: </strong>
            <br /> {props.info.Ports.IP}{' '}
          </li>
          <hr />
          <li>
            <strong>PrivatePort: </strong>
            <br /> {props.info.Ports.PrivatePort}{' '}
          </li>
          <hr />
          <li>
            <strong>PublicPort: </strong>
            <br /> {props.info.Ports.PublicPort}{' '}
          </li>
          <hr />
          <li>
            <strong>Type: </strong>
            <br /> {props.info.Ports.Type}{' '}
          </li>
        </ul>
      </div>
    );
  }
  //else return container information without Ports info
  return (
    <div id={props.id} className='container'>
      {/* Display container information*/}
      <div className='containerInfo'>
        <p>
          {' '}
          <strong>Name: </strong>
          <br /> {props.info.Name}
        </p>
        <hr />
        <p>
          <strong>ContainerID: </strong>
          <br /> {props.info.Id}
        </p>
        <hr />
        <p>
          <strong>Image: </strong>
          <br /> {props.info.Image}
        </p>
        <hr />
        <p>
          <strong>Activity: </strong>
          <br /> {props.info.State}
        </p>
        <div className='containerButtons'>
          {/* <button className= 'button' onClick={ () => ConnectContainers(props.info.Name)}>Connect</button> */}
          <button
            className='innerButton'
            onClick={() =>
              DisconnectContainer(
                props.info.Name,
                props.network,
                props.setContainers,
                props.setNetworks,
                props.isLoading
              )
            }>
            Disconnect
          </button>
        </div>
      </div>
    </div>
  );
};

export default ContainerDisplay;
