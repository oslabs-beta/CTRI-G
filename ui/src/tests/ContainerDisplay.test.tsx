import { render, screen, fireEvent } from '@testing-library/react';

import '@testing-library/jest-dom';
import ContainerDisplay from '../components/ContainerDisplay';
import { NetworkInfo } from '../interfaces/interfaces';
import { create } from 'react-test-renderer';

jest.mock('@docker/extension-api-client', () => ({
  createDockerDesktopClient: jest.fn().mockReturnValue({
    docker: {
      clie: {
        exec: jest.fn(),
      },
    },
    desktopUI: {
      toast: {
        error: jest.fn(),
        success: jest.fn(),
        warning: jest.fn(),
      },
    },
  }),
}));

describe('Container component unit tests', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.resetModules();
  });

  const portItem = {
    IP: '0.0.0.0',
    PrivatePort: '9090',
    PublicPort: '9090',
    Type: 'tcp',
  };
  const containerInfoWithPorts = {
    Name: 'prometheus',
    Id: '123456',
    Image: 'image',
    State: 'running',
    Networks: 'containerwatch_containerwatch-desktop-extension_default',
    Ports: [portItem],
  };

  const containerInfoWithoutPorts = {
    Name: 'prometheus',
    Id: '123456',
    Image: 'image',
    State: 'running',
    Networks: 'containerwatch_containerwatch-desktop-extension_default',
    Ports: [],
  };

  const networkInfo: NetworkInfo = {
    Driver: 'bridge',
    Name: 'bridge',
    ID: '1234',
    Containers: [containerInfoWithPorts.Name],
    IPv4Address: ['0.0.0.0'],
  };

  test('renders the container display component without port information', () => {
    render(
      <ContainerDisplay
        id={containerInfoWithoutPorts.Id}
        info={containerInfoWithoutPorts}
        network={networkInfo}
        containerIndex={0}
      />
    );

    const element = screen.getByText(/prometheus/i);

    expect(element).toBeInTheDocument();
  });

  test('does not render any port information if there is no port info to be rendered', () => {
    render(
      <ContainerDisplay
        id={containerInfoWithoutPorts.Id}
        info={containerInfoWithoutPorts}
        network={networkInfo}
        containerIndex={0}
      />
    );

    const element = screen.queryByText(/PrivatePort/i);

    expect(element).not.toBeInTheDocument();
  });

  test('renders the container display component with port information', () => {
    render(
      <ContainerDisplay
        id={containerInfoWithPorts.Id}
        info={containerInfoWithPorts}
        network={networkInfo}
        containerIndex={0}
      />
    );

    const portElement = screen.getByText(/Private Ports/);

    const containerElement = screen.getByText(/image/);

    expect(portElement).toBeInTheDocument();
    expect(containerElement).toBeInTheDocument();
  });

  test('renders formModal if the connect to networks button is clicked', () => {
    render(
      <ContainerDisplay
        id={containerInfoWithoutPorts.Id}
        info={containerInfoWithoutPorts}
        network={networkInfo}
        containerIndex={0}
      />
    );
    const buttonElement = screen.getByRole('button', {
      name: 'Connect to Networks',
    });
    fireEvent.click(buttonElement);

    const outputElement = screen.getByText(/Container to a Network/);
    expect(outputElement).toBeVisible();
  });
  test('Does not render formModal if the connect to networks button is not clicked', () => {
    render(
      <ContainerDisplay
        id={containerInfoWithoutPorts.Id}
        info={containerInfoWithoutPorts}
        network={networkInfo}
        containerIndex={0}
      />
    );

    const outputElement = screen.queryByText(/Container to a Network/);
    expect(outputElement).not.toBeInTheDocument();
  });

  test('renders ContainerDisplay', () => {
    const containerDisplay = create(
      <ContainerDisplay
        id={containerInfoWithoutPorts.Id}
        info={containerInfoWithoutPorts}
        network={networkInfo}
        containerIndex={0}
      />
    ).toJSON();
    expect(containerDisplay).toMatchSnapshot();
  });
});
