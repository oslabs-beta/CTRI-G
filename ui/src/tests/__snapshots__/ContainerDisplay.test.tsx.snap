// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container component unit tests renders ContainerDisplay 1`] = `
<div
  className="container"
  id="123456"
>
  <div
    className="containerHeader"
  >
    <p>
       
      <strong>
        Container: 
      </strong>
      <span />
       
      prometheus
    </p>
  </div>
  <div
    className="containerInfoContainer"
  >
    <div
      className="containerInfo withPorts"
    >
      <hr
        className="tophr"
      />
      <p
        className="clipText"
      >
        <strong>
          ContainerID: 
        </strong>
        <br />
         
        123456
      </p>
      <hr />
      <p
        className="clipText"
      >
        <strong>
          Image: 
        </strong>
        <br />
         
        image
      </p>
      <hr />
      <p>
        <strong>
          Activity: 
        </strong>
        <br />
         
        running
      </p>
      <hr
        className="bottomhr"
      />
    </div>
    <ul
      className="portInfo"
    >
      <li>
        <strong>
          Type: 
        </strong>
         
      </li>
      <hr />
      <li>
        <strong>
          IPv4Address: 
        </strong>
        0.0.0.0
      </li>
      <hr />
      <li>
        <strong>
          Published Ports: 
        </strong>
        <em>
          null
        </em>
      </li>
      <hr />
      <li>
        <strong>
          Private Ports: 
        </strong>
        <em>
          null
        </em>
      </li>
      <li>
        <button
          className="innerButton editButton"
          onClick={[Function]}
        >
          <p>
            Edit Ports
          </p>
        </button>
      </li>
    </ul>
  </div>
  <div
    className="footerContainer"
  >
    <div
      className="containerButtons"
    >
      <button
        className="innerButton"
        onClick={[Function]}
      >
        Disconnect
      </button>
      <button
        className="innerButton netconnect"
        onClick={[Function]}
      >
        Connect to Networks
      </button>
    </div>
  </div>
</div>
`;
